witness/golang/internal/witness/witness.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package witness is designed to make sure the checkpoints of verifiable logs
+// are consistent and store/serve/sign them if so.  It is expected that a separate
+// feeder component would be responsible for the actual interaction with logs.
+package witness
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/logverifier"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// Chkpt consists of the size of a checkpoint and its raw format.
+type Chkpt struct {
+	Size uint64
+	Raw  []byte
+}
+
+// ChkptStorage is an interface for reading and writing checkpoints.
+type ChkptStorage interface {
+	// GetLatest returns the latest checkpoint for a given log.
+	GetLatest(logID string) (*Chkpt, error)
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log with
+	// latestSize as the size of its largest stored checkpoint.
```

Comment:
It would be nice to reword this to emphasise that this is a conditional write, i.e. the write will happen only if the latest stored checkpoint size is `latestSize`.

Reply:
Done.